File: _API_TESTING_GHERKIN_/DEMOAPP001_TYPESCRIPT_CYPRESS/src/screenplay/tasks/SendGetRequest.ts
import type { Task } from "../core/types";
import type { Actor } from "../actors/Actor";
import { CallAnApi } from "../abilities/CallAnApi";
import { LAST_RESPONSE } from "../support/memory-keys";

export class SendGetRequest implements Task {
  private constructor(
    private readonly path: string,
    private readonly options?: Partial<Cypress.RequestOptions>
  ) {}

  static to(path: string, options?: Partial<Cypress.RequestOptions>): SendGetRequest {
    return new SendGetRequest(path, options);
  }

  performAs(actor: Actor): Cypress.Chainable<unknown> {
    const api = actor.abilityTo(CallAnApi);
    return api.get(this.path, this.options).then((response: Cypress.Response<unknown>) => {
      actor.remember(LAST_RESPONSE, response);
      return response as unknown;
    });
  }
}

File: _API_TESTING_GHERKIN_/DEMOAPP002_CSHARP_PLAYWRIGHT/TokenParserTests/Screenplay/Tasks/SendGetRequest.cs
using System.Web;
using TokenParserTests.Screenplay.Abilities;
using TokenParserTests.Screenplay.Support;

namespace TokenParserTests.Screenplay.Tasks;

public sealed class SendGetRequest : ITask
{
    private readonly string _endpoint;
    private readonly IReadOnlyDictionary<string, string>? _query;

    private SendGetRequest(string endpoint, IReadOnlyDictionary<string, string>? query)
    {
        _endpoint = endpoint;
        _query = query;
    }

    public static SendGetRequest To(string endpoint, IReadOnlyDictionary<string, string>? query = null) =>
        new(endpoint, query);

    public async Task PerformAsAsync(Actor actor, CancellationToken cancellationToken = default)
    {
        var callAnApi = actor.AbilityTo<CallAnApi>();

        var requestUri = BuildUri(_endpoint, _query);
        var response = await callAnApi.GetAsync(requestUri, cancellationToken).ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        actor.Remember(MemoryKeys.LastResponseStatus, response.StatusCode);
        actor.Remember(MemoryKeys.LastResponseBody, body);
    }

    private static string BuildUri(string endpoint, IReadOnlyDictionary<string, string>? query)
    {
        if (query == null || query.Count == 0)
        {
            return endpoint;
        }

        var builder = HttpUtility.ParseQueryString(string.Empty);
        foreach (var kvp in query)
        {
            builder[kvp.Key] = kvp.Value;
        }

        return $"{endpoint}?{builder}";
    }
}
